client/core/device.go

```diff
@@ -38,6 +40,18 @@ type Device struct {
 	lock         sync.Mutex
 }
 
+func (d *Device) UpdateBy(v *Device) {
+	d.setDeviceID(v.DeviceID())
+	// foundByIP can be overwritten only when it is set.
+	if v.foundByIP.Load() != "" {
```

Comment:
Technically, the value of `v.foundByIP` can be different on lines 46 and 47, if the scheduler changes threads after line 46 but before 47. For correctness, you should save it to a local variable and then only use that.

Reply:
fixed